algorithms/mopt/mopt_optimizer.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/mopt/mopt_optimizer.hpp"
+
+#include <random>
+#include <vector>
+
+#include "fuzzuf/mutator/havoc_case.hpp"
+#include "fuzzuf/optimizer/keys.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::optimizer {
+
+using fuzzuf::mutator::INSERT_AEXTRA;
+using fuzzuf::mutator::INSERT_EXTRA;
+using fuzzuf::mutator::OVERWRITE_WITH_AEXTRA;
+using fuzzuf::mutator::OVERWRITE_WITH_EXTRA;
+
+MOptParticle::MOptParticle() {}
+MOptParticle::~MOptParticle() {}
+
+MOptOptimizer::MOptOptimizer() {
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::NewTestcases, (u64)0);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::HavocOperatorFinds);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::SelectedCaseHistogram);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::LastSpliceCycle, (u32)0);
+
+  for (auto& p : swarm) {
+    for (auto& pos : p.position)
+      pos = fuzzuf::utils::random::Random<double>(min_position, max_position);
+    p.velocity.fill(0);
+  }
+
+  idx = 0;
+  time = 0;
+
+  UpdateInertia();
+}
+
+MOptOptimizer::~MOptOptimizer() {}
+
+void MOptOptimizer::SetScore(size_t i, double score) {
+  auto& p = swarm[idx];
+  p.fitness[i] = score;
+}
+
+void MOptOptimizer::UpdateLocalBest() {
+  auto& p = swarm[idx];
+
+  if (unlikely(time == 0)) {
+    for (size_t i = 0; i < p.fitness.size(); i++) {
+      p.best_fitness[i] = p.fitness[i];
+      p.best_position[i] = p.position[i];
+    }
+  }
```

Comment:
I think, if you do this, the following loop should be included in `if (time ==0) {} else { ... }`.
But personally I prefer removing this block and initializing p.best_fitness[i] with -1 or -∞ or something

Reply:
Also, it seems `time` is incremented nowhere.

Reply:
Considering that UpdateLocalBest and UpdateGlobalBest both want to have initialization while their initialization takes place at different times, it would be dangerous to depend on the sole value `time`.
So I recommend you to remove the usage of `time`.